### PR TITLE
New high level API and general rework

### PR DESCRIPTION
**Pull request by mnottoli**
The old high level Fortran APIs were not properly done, and the C ones were missing. This pull request addresses these two problems and brings in some additional content to make future development easier.

This is a list of the content:

- New model nonspecific APIs in Fortran and C. There is a set of functions to set up, fill the guesses, solve the linear systems and compute energies and forces. To circumvent interface problems, all the electrostatic properties are now stored in a container (`ddx_electrostatics_type`)
- New monolithic solver routine in Fortran and C. A new version of `ddsolve` has been implemented, this solves the linear systems and computes the energy and (if requested) the forces. Now, it uses the electrostatic properties container and optional arguments to avoid unnecessary allocations. A similar version is also implemented in C.
- Consistent naming. All the APIs have consistent names in Python, C and Fortran.
- Moved the deprecated content (only used by the tests) to the module ddx_legacy.
- Added a new `ddinit` which is a wrapper around `allocate_model` and uses optional arguments. This makes it possible to add future extensions without breaking retro-compatibility with software that it is not aware of the new arguments (e.g. in case of new models / new functionality that requires some settings).

### CODE REVIEW

**Comment by mnottoli** [comment]
This is the last of the three pull requests. @mfherbst, I will send you further information.

All the tests (Python, Fortran, Psi4) are passing.

This builds up on #148. 
To close this, we can do the following. After #148 gets merged I would suggest creating a new local branch, cherry-picking all the relevant commit from this PR (from [a444609](https://github.com/ddsolvation/ddX/pull/149/commits/a444609328cdb23b9d751295bedda304152f9e04) on) and then rebasing on the new main. After that, a git push --force can be used to overwrite the content of this PR, with the rebased PR. This avoids creating a new PR.

**Review by mfherbst — COMMENTED** [review]
Looks good, mainly merge conflicts need to be resolved.

File: examples/run_ddx.c [hunk 1]
```diff
@@ -78,52 +78,69 @@ int main() {
...
-  ddx_pcm_guess(model, state, error);
-  ddx_pcm_solve(model, state, tol, error);
+  if (fabs(eref - energy) > 1e-6) {
+    printf("Large deviation:   %15.9g\n", eref - energy);
+    ret = 1;
+  }
+
+  //
+  // ... or solve the PCM problem step by step
+  //
```
[the remark this answers is not in the document]
Comment (mnottoli) [reply examples/run_ddx.c]: I split the example in two.

File: src/ddx.h [hunk 2]
```diff
@@ -252,6 +271,86 @@ void ddx_get_xi(const void* state, const void* ddx, int ncav, double* xi);
...
+ *  \param state   DDX state
+ *  \param electrostatics DDX electrostatic properties container
+ *  \param nsph    Number of cavity spheres
+ *  \param forces  Output force array (3, nsph) in column-major order
+ *  \param error   DDX error
+ */
+void ddx_solvation_force_terms(const void* ddx, void* state, void* electrostatics,
+                               int nsph, double* forces, void* error);
+
+double ddx_ddrun(const void* ddx, void* state, void* electrostatics, int nbasis,
```
[the remark this answers is not in the document]
Comment (mnottoli) [reply src/ddx.h:360]: Documentation added.

File: src/ddx.h [hunk 3]
```diff
@@ -60,6 +60,25 @@ void ddx_get_error_message(const void* error, char* message, int maxlen);
...
+                                  const double* multipoles,
+                                  void* electrostatics, void* error);
+
+/** Deallocate the ddX electrostatics object. */
+void ddx_deallocate_electrostatics(void* electrostatics, void* error);
+
+///@}
+
+/** \name Allocate and manage the ddx model object */
+///@{
```
[the remark this answers is not in the document]
Comment (mnottoli) [reply src/ddx.h]: Removed.

**Comment by mnottoli** [comment]
Rebase done, ready to be merged.

**Pull request merged**